Libraries/LibCore/Account.cpp

```diff
@@ -0,0 +1,181 @@
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <AK/Base64.h>
+#include <AK/Random.h>
+#include <LibCore/Account.h>
+#include <LibCore/File.h>
+#include <grp.h>
+#include <pwd.h>
+#include <string.h>
+#include <unistd.h>
+
+namespace Core {
+
+static String get_salt()
+{
+    char random_data[12];
+    AK::fill_with_random(random_data, sizeof(random_data));
+
+    StringBuilder builder;
+    builder.append("$5$");
+    builder.append(encode_base64(ReadonlyBytes(random_data, sizeof(random_data))));
+
+    return builder.build();
+}
+
+static Vector<gid_t> get_gids(const StringView& username)
+{
+    Vector<gid_t> extra_gids;
+    for (auto* group = getgrent(); group; group = getgrent()) {
+        for (size_t i = 0; group->gr_mem[i]; ++i) {
+            if (username == group->gr_mem[i])
+                extra_gids.append(group->gr_gid);
+        }
+    }
+    endgrent();
+    return extra_gids;
+}
+
+Result<Account, String> Account::get(const StringView& username)
+{
+    String username_with_terminator(username);
+
+    struct passwd* pwd = nullptr;
+    pwd = getpwnam(username_with_terminator.characters());
+    if (!pwd)
+        return String("No such user");
+
+    Account account(pwd, get_gids(pwd->pw_name));
+    endpwent();
+    return account;
+}
+
+Result<Account, String> Account::get(uid_t uid)
+{
+    struct passwd* pwd = nullptr;
+    pwd = getpwuid(uid);
+    if (!pwd)
+        return String("No such user");
+
+    Account account(pwd, get_gids(pwd->pw_name));
+    endpwent();
+    return account;
+}
+
+bool Account::authenticate(const StringView& password) const
+{
+    if (m_passwd == "")
```

Comment:
```suggestion
    if (m_passwd.is_empty())
```

Also, does the empty password have a special meaning?

Reply:
yes. an empty password means no password is required to authenticate as the user.

Reply:
Can the user not have an actual empty password?

Or, perhaps this is an empty *encrypted* password, not the password itself? In that case, can we maybe use null instead of an empty string?

In any case, this needs a comment or two 🙂

Reply:
an actual empty password would still be hashed. The use of an empty password field to mean no password is a unix thing.

from linux's `passwd(5)`:
> The encrypted password field may be empty, in which case no password is required to authenticate as the specified login name.

If we were to use null, we would have to convert to and from null when reading and writing the field.